### PR TITLE
Add `comment_status` field to quick edit

### PR DESCRIPTION
**Pull request by oandregal**
Part of https://github.com/WordPress/gutenberg/issues/55101 https://github.com/WordPress/gutenberg/issues/59745

## What?

Adds `comment_status` (Discussion) field to quick edit.

## How?

- Even if it seems a boolean, the data source ([REST API](https://developer.wordpress.org/rest-api/reference/pages/#create-a-page)) models it as a `text` field with two elements (open & close). That's how the field is added as well.
- Introduces a `editAs` function that enables the consumer to configure the Edit function.

## Testing Instructions

- Visit the Pages page in the site editor with the quick edit experiment enabled. 
- Select one item and change its "Discussion" status.
- Go to the editor and verify that it was changed.

### CODE REVIEW

File: packages/edit-site/src/components/post-fields/index.js [hunk 1]
```diff
@@ -344,6 +344,16 @@ function usePostFields( viewType ) {
 					return <time>{ getFormattedDate( item.date ) }</time>;
 				},
 			},
+			{
+				id: 'comment_status',
+				label: __( 'Discussion' ),
+				type: 'text',
+				editAs: 'radio',
```
Comment (oandregal) [inline packages/edit-site/src/components/post-fields/index.js]: The REST API doesn't define any boolean data type, so we'll have to wait to implement that.

I took this as an opportunity to explore how to provide a different edit component (radio) than the default (select), hence the `editAs` prop. I also considered the following alternatives:

1. We only provide `select` as default and consumer can provide its own `edit` function. This seems too basic.
2. Alternative names for the prop: `edit`, overloading the custom `edit` function the consumer can provide — could be nice.

File: packages/dataviews/src/types.ts [hunk 2]
```diff
@@ -94,6 +94,11 @@ export type Field< Item > = {
 	 */
 	Edit?: ComponentType< DataFormControlProps< Item > >;
 
+	/**
+	 * Optional config for editing the field.
+	 */
+	editAs?: 'radio';
```
Comment (youknowriad) [inline packages/dataviews/src/types.ts]: I do think we need a config like that. Maybe I would name it "control". 

I think this also means that like "field types" we need a folder of "controls" and we already have three of them (NumberControl, TextControl and SelectControl). I believe these should be treated similarly to RadioControl.

File: packages/dataviews/src/field-types/text.tsx [hunk 3]
```diff
@@ -49,6 +53,18 @@ function Edit< Item >( {
 		[ id, onChange ]
 	);
 
+	if ( field.elements && field.editAs === 'radio' ) {
```
Comment (youknowriad) [inline packages/dataviews/src/field-types/text.tsx]: I think we need to remove all these conditions and have a folder for control and something like:

```
const Control = getFieldControl( field )

<Control />
```

**Review by youknowriad — COMMENTED** [review]
Great work

File: packages/dataviews/src/types.ts (hunk 2 above)
Comment (oandregal) [reply packages/dataviews/src/types.ts]: Given there's a `Edit` prop to provide a custom edit function, how about `editControl` so it's connected by the prefix?

My thinking is that "controls" seem too open-ended: it seems you can list any control and it'll be used. I see it more as a "edit configuration" if that makes sense.

Comment (youknowriad) [reply packages/dataviews/src/types.ts]: The question I'm wondering though is whether this should be part of the `field` definition or the `form` object.

Comment (youknowriad) [reply packages/dataviews/src/types.ts]: It is "edit" configuration yes, but I don't like the name "editAs" personally. I think the question also is whether we see folks registering new "controls" without necessarily changing the "type" of the field. And the answer for me seems "yes" given that the "type" controls more than just "edit", I want to use the "integer" sorting/validation... but I want to build a custom control for my field.

File: packages/edit-site/src/components/post-fields/index.js [hunk 4]
```diff
@@ -344,6 +344,28 @@ function usePostFields( viewType ) {
 					return <time>{ getFormattedDate( item.date ) }</time>;
 				},
 			},
+			{
+				id: 'comment_status',
```
Comment (ntsekouras) [inline packages/edit-site/src/components/post-fields/index.js:349]: We should disable sorting here, as it's not supported.

Comment (ntsekouras) [reply packages/edit-site/src/components/post-fields/index.js:349]: Additionally I think filtering is not supported by the REST API too..

File: packages/edit-site/src/components/post-fields/index.js (hunk 1 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/post-fields/index.js]: >The REST API doesn't define any boolean data type, so we'll have to wait to implement that.

What do you mean with this?

Comment (oandregal) [reply packages/edit-site/src/components/post-fields/index.js]: That all data in https://developer.wordpress.org/rest-api/reference/pages/#schema is string, integer, or object.

File: packages/dataviews/src/types.ts (hunk 2 above)
Comment (ntsekouras) [reply packages/dataviews/src/types.ts]: How about changing the `Edit` signature to also support a predefined set of controls?
So it would be `Edit: 'radio'`.

File: packages/edit-site/src/components/post-fields/index.js (hunk 1 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/post-fields/index.js]: Yeah, for pages it doesn't. But what will we have to wait to implement?

File: packages/dataviews/src/types.ts (hunk 2 above)
Comment (youknowriad) [reply packages/dataviews/src/types.ts]: > How about changing the Edit signature to also support a predefined set of controls?
So it would be Edit: 'radio'.

I like that but I would still prefer "control" over "edit".

File: packages/edit-site/src/components/post-fields/index.js (hunk 1 above)
Comment (oandregal) [reply packages/edit-site/src/components/post-fields/index.js]: The boolean type field until it's in use.

File: packages/edit-site/src/components/post-fields/index.js (hunk 4 above)
Comment (oandregal) [reply packages/edit-site/src/components/post-fields/index.js:349]: Oh, thanks! I actually forgot all fields are part of DataViews, even if hidden. Assumed it worked like DataForm and you had to add them to the allowed list. Fixed at fcb8cc4d50408ccf07baf43ceefb42d2fe4a0ad0

File: packages/dataviews/src/types.ts (hunk 2 above)
Comment (oandregal) [reply packages/dataviews/src/types.ts]: I reworked this to overload `Edit`, this was my [alternative option](https://github.com/WordPress/gutenberg/pull/64370#discussion_r1709223705) as well.

**Comment by oandregal** [comment]
I pushed an initial version that introduces controls (only `radio`) to gather feedback while I further refactor the field type definitions to use controls.

File: packages/dataviews/src/components/dataform-controls/radio.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/**
```
Comment (youknowriad) [inline packages/dataviews/src/components/dataform-controls/radio.tsx:1]: Personally, I would move the "text", "radio", "select" and "integer control to. `src/controls` or something like that. But can be done in a separate PR.

File: packages/dataviews/src/types.ts [hunk 6]
```diff
@@ -53,6 +53,26 @@ export type ValidationContext = {
 	elements?: Option[];
 };
 
+/**
+ * An abstract interface for Field based on the field type.
+ */
+export type FieldTypeDefinition< Item > = {
```
Comment (youknowriad) [inline packages/dataviews/src/types.ts:59]: 👍

One thing I'm wondering is that "field types" shouldn't depend on "Item" they should work for any item value of the defined type. So I think we have somethings wrong in the abstraction. `sort, isValid and Edit` should probably not receive `items` at all and just receive values. But it's also something that should be solved in a separate PR.

**Comment by youknowriad** [comment]
While testing this, I noticed that something regressed recently when it comes to long columns (long titles)

Screenshot 2024-08-09 at 9 49 41 AM

cc @jameskoster if you're aware of a related change.

**Review by youknowriad — APPROVED** [review]
There's a couple of follow-ups I think we should do but this is testing well for me.

**Comment by oandregal** [comment]
> There's a couple of follow-ups I think we should do but this is testing well for me.

Ok, I'll merge as it is then. I'm going to set this to auto-merge and focus a bit on clearing my queue — I don't expect to come back to these follow-ups today.
